testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ee/concurrent/ManagedExecutorServiceMetricsTestCase.java

```diff
@@ -0,0 +1,248 @@
+            } catch (IOException e) {
+                Assert.fail();
+            }
+        }).get();
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.ACTIVE_THREAD_COUNT) == 0);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.COMPLETED_TASK_COUNT) == 2);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.CURRENT_QUEUE_SIZE) == 0);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.HUNG_THREAD_COUNT) == 0);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.MAX_THREAD_COUNT) == 2);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.TASK_COUNT) == 2);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.THREAD_COUNT) == 2);
+
+        // exclusive testing of queue size stat
+        CyclicBarrier barrier1 = new CyclicBarrier(3);
+        CyclicBarrier barrier2 = new CyclicBarrier(3);
+        executorService.execute(() -> {
+            logger.info("Executing task #3.1");
+            try {
+                barrier1.await();
+                barrier2.await();
+            } catch (Exception e) {
+                Assert.fail();
+            }
+        });
+        executorService.execute(() -> {
+            logger.info("Executing task #3.2");
+            try {
+                barrier1.await();
+                barrier2.await();
+            } catch (Exception e) {
+                Assert.fail();
+            }
+        });
+        barrier1.await();
+        // 2 tasks running, executing a 3rd should place it in queue,
+        // cause when core threads is reached executor always prefers queueing than creating a new thread
+        executorService.execute(() -> {
+            logger.info("Executing task #3.3");
+        });
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.CURRENT_QUEUE_SIZE) == 1);
+        // resume tasks running and confirm expected idle stats
+        barrier2.await();
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.ACTIVE_THREAD_COUNT) == 0);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.COMPLETED_TASK_COUNT) == 5);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.CURRENT_QUEUE_SIZE) == 0);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.MAX_THREAD_COUNT) == 2);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.TASK_COUNT) == 5);
+        Assert.assertTrue(readStatsAttribute(pathAddress, ManagedExecutorServiceMetricsAttributes.THREAD_COUNT) == 2);
```

Comment:
All these would be racy as we should only be checking once we no all 3 threads have completed.

Reply:
You are correct, there was a race between tasks completing and getting stats. Honestly I don't recall why I used execute() in this final test, anyway now fixed by replacing execute() with submit(), and then using each future.get() to ensure all tasks complete before checking for idle stats.